audits/trading/202404-threat-model-immutable-signed-zone-v2.md

```diff
@@ -2,14 +2,214 @@
+
+* It requires orders to be known by the orderbook prior to fulfilment
+* It allows the orderbook to refuse fulfilment of an order at its discretion (e.g. for gasless cancellations, or compromised collections)
+* It allows for arbitrary logic to be defined off-chain and enforced on-chain (e.g. to enforce fees)
+
+## Threat Model Scope
+
+The threat model is limited to the following Solidity files at GitHash [TBD]:
+
+* [ImmutableSignedZoneV2.sol [TBD]]()
+* [SIP5EventsAndErrors.sol [TBD]]()
+* [SIP5Interface [TBD]]()
+* [SIP6EventsAndErrors [TBD]]()
+* [SIP6Interface [TBD]]()
+* [SIP7EventsAndErrors [TBD]]()
+* [SIP7Interface [TBD]]()
+
+## Background
+
+Seaport: [Documentation](https://github.com/ProjectOpenSea/seaport/blob/20b84b94755ab4fcdd88735d5f8f1f578e07924a/docs/SeaportDocumentation.md) for v1.6 but relevant for v1.5 which Immutable's Seaport fork is based on.
+
+SIP-7: [Documentation [TBD]](https://github.com/ProjectOpenSea/SIPs/blob/main/SIPS/sip-7.md).
 
 ## Architecture
 
+### Top Level Architecture
+
+The trading system on the Immutable platform is shown in the diagram below.
+
+```mermaid
+flowchart LR
+  client[Client] <-- 1. POST .../fulfillment-data ---> ob[Immutable Off-Chain\nOrderbook]
+  client -- 2. fulfillAdvancedOrder ---> seaport[ImmutableSeaport.sol]
+  seaport -- 3a. transferFrom --> erc20[IERC20.sol]
+  seaport -- 3b. transferFrom --> erc721[IERC721.sol]
+  seaport -- 3c. safeTransferFrom --> erc1155[IERC1155.sol]
+  seaport -- 4. validateOrder --> Zone
+  subgraph Zone
+    direction TB
+    zone[ImmutableSignedZoneV2.sol] --> AccessControlEnumerable.sol
+  end
+```
+
+The sequence of events is as follows:
+
+1. The client makes a HTTP `POST .../fulfillment-data` request to the Immutable Orderbook, which will construct and sign an `extraData` payload to return to the client
+2. The client calls `fulfillAdvancedOrder` or `fulfillAvailableAdavancedOrders` on `ImmutableSeaport.sol` to fulfill an order
+3. `ImmutableSeaport.sol` executes the fufilment by transferring items between parties
+4. `ImmutableSeaport.sol` calls `validateOrder` on `ImmutableSignedZoneV2.sol`, passing it the fulfilment execution details as well as the `extraData` parameter
+   1. `ImmutableSignedZoneV2.sol` validates the fulfilment execution details using the `extraData` payload, reverting if expectations are not met
+
+### Immutable Signed Zone (v2) Design
+
+Immutable Signed Zone (v2) is an implementation of the SIP-7 specification, supporting substandards 3, 4 and 6.
+
+This section explains the security relevant aspects of the Immutable Signed Zone (v2) contract beyond the scope of the SIP-7 specification.
+
+#### Roles
+
+The Immutable Signed Zone (v2) supports the following roles:
+
+* Default Admin
+  * Creates and removes other administrators
+  * First admin is assigned to the `address owner` param on the `constructor`
```

Comment:
Is this owner related to linking the contract to an account in Immutable Hub? If so, it would be good to switch to what is being done in Assets / what they are moving towards, where the DEFAULT_ADMIN_ROLE just deals with role administration, and there is a HUB_OWNER_ROLE.

Reply:
Our contract is intended to be deployed and managed by Immutable exclusively. Hub is not involved in this management.

Reply:
The owner will be set to the multi-sig